### PR TITLE
Phase - 1 Completion

### PR DESCRIPTION
**Pull request by ccpandhare**
This commit adds support for Node.js, addSteps is created and accepts input normally and in JSON format. Since this is not yet fully functional -- only addSteps and loadImage work as of now, I have maintained this as a different file. ImageSequencerNode.js and ModulesNode.js. Multiple image support has been added. Now the images reside in `sequencer.images`. Every image has its own `steps` object and every image has different steps.

demonstration of addSteps can be found at the [Development Wiki](https://github.com/publiclab/image-sequencer/wiki/Development).

The modules NdviRed, DoNothing, GreenChannel have been modifed to work with this new framework as of now.

./index.js demonstrates what all can be done.

Simply `node -r ./index` for testing purposes.

I know this is a very short description for a huge number of changes. I will write a detailed explanation shortly.

### CODE REVIEW

**Comment by jywarren** [comment]
This sounds great -- what's your schedule for reunifying the two versions,
and do you think you want to merge this in before that's complete?

**Comment by ccpandhare** [comment]
I will not be reunifying the two versions, rather, the older one will be replaced. the only reason I have kept it is because the newer one doesn't have all the functionality yet. I shall do the replacing by the end of Phase-1 as defined in the [Planning Issue](https://github.com/publiclab/image-sequencer/issues/23).

I think this should be merged into a separate branch - "development" for testing purposes, what are your views on this?

**Comment by jywarren** [comment]
You can open PRs against other PRs, actually - so that would be fine. I
guess as soon as possible if we can get to a point where this is merged
into master, and the new replaces the old, that's ideal, because having so
many tangled big PRs is not great for newcomers. But i know it's just an
element of starting up a big project and won't be happening as much as the
code matures. We def. want to be in a place soon where we can be making
smaller, self-contained, single-function PRs that are easy to read through
and test!

**Comment by ccpandhare** [comment]
Okay! Wow. I hadn't really gone that far with my thinking. Thanks for this
insight!

**Comment by jywarren** [comment]
Cool, what's the status on this now? Is `output.txt` required as part of testing?

**Comment by ccpandhare** [comment]
Oh no, not at all. I shall remove it.
Also, I am currently modifying the Readme and making the code more legible and clean.
Then we could merge this in!

**Comment by jywarren** [comment]
Super. Exciting!

File: README.md [hunk 1]
```diff
@@ -28,9 +28,78 @@ Examples:
 * [Basic example](https://jywarren.github.io/image-sequencer/)
 * [NDVI example](https://jywarren.github.io/image-sequencer/examples/ndvi/) - related to [Infragram.org](http://infragram.org)
 
+Using the library:
+The Image Sequencer Library exports a function ImageSequencer which initializes a sequencer.
+```js
```
Comment (jywarren) [inline README.md]: Sorry to be picky here but I think clear empty lines between paragraphs is a good thing for readability -- especially when on a command line interface and reading the raw Markdown. Thanks!

File: README.md [hunk 2]
```diff
@@ -28,9 +28,78 @@ Examples:
...
+Steps can be removed using the `removeSteps` method. It accepts `image` and `index` as parameters.
+Either, both, or none of them can be an array. JSON input is also accepted. This method automatically triggers the `run` method as a callback.
+```js
+sequencer.removeSteps("image",[steps]);
+```
+```js
+sequencer.removeSteps("image",step);
+```
+```js
+sequencer.removeSteps({
```
Comment (jywarren) [inline README.md]: All this is really great. I actually hope your tests are similarly laid out -- tests can, for example, have long text descriptions in the spec name, so you can explain things at this level of detail. Then your API will match your README!

File: README.md [hunk 3]
```diff
@@ -28,9 +28,78 @@ Examples:
...
+  image: [steps],
+  ...
+});
+```
+
+
 ## Contributing
 
-Happily accepting pull requests; to edit the core library, modify files in `/src/`. To build, run `npm install` and `grunt build`. 
+Happily accepting pull requests; to edit the core library, modify files in `/src/`. To build, run `npm install` and `grunt build`.
```
Comment (jywarren) [inline README.md]: Note something like `all files from /src/ will be compiled into /dist/.....`

File: index.js [hunk 4]
```diff
@@ -2,7 +2,8 @@ console.log('\x1b[31m%s\x1b[0m',"This is the output of the module");
 require('./src/ImageSequencer');
 sequencer = ImageSequencer();
 sequencer.loadImages({red:'examples/red.jpg'},function(){
-  sequencer.addSteps(['do-nothing','do-nothing-pix','do-nothing']);
+  sequencer.addSteps(['do-nothing','do-nothing-pix','do-nothing','do-nothing']);
```
Comment (jywarren) [inline index.js]: Is this mixed in from your testing? Maybe remove from this PR?

File: README.md (hunk 1 above)
Comment (ccpandhare) [inline README.md]: Okay! I will do that!

File: README.md (hunk 2 above)
Comment (ccpandhare) [inline README.md]: Yes the tests are laid out similarly. Tests aren't complete yet. A few still remain.
I actually don't understand you clear... Could you explain this a bit, please?

File: README.md (hunk 3 above)
Comment (ccpandhare) [inline README.md]: Okay, sure!
Also, should we keep the "Contributing" header in the Readme any longer?
Just that the Readme is getting really long!

Comment (jywarren) [inline README.md]: I think it's ok for now. Thanks!

File: README.md (hunk 2 above)
Comment (jywarren) [inline README.md]: Just saying that I like how clear these are, and we should try to make the tests just as clear -- and if they are in the same order and everything, we can even link to the tests as a "provably working" copy of this set of commands. Like: "see this actually running, in context", you know? 

And as to the naming, you can reuse this text -- so on line 105 here, you can add a matching test:

```js
test('Steps can be removed using the `removeSteps` method. It accepts `image` and `index` as parameters.', function (t) {
```

Make sense?

**Comment by jywarren** [comment]
This is good to go now, right? Go ahead and merge it if so!

My final comments were related to the relation between docs and tests:

***

Just saying that I like how clear these are, and we should try to make the tests just as clear -- and if they are in the same order and everything, we can even link to the tests as a "provably working" copy of this set of commands. Like: "see this actually running, in context", you know? 

And as to the naming, you can reuse this text -- so on line 105 here, you can add a matching test:

```js
test('Steps can be removed using the `removeSteps` method. It accepts `image` and `index` as parameters.', function (t) {
```

Make sense?

**Comment by ccpandhare** [comment]
Ohh I see. Okay. I agree. Will do that. But since the tests aren't complete, I think that could all go into a separate PR.

I am afraid I don't have merge rights. Could you merge the changes for me?
Thanks.

**Comment by jywarren** [comment]
That's fine, sounds good. Oh, and sorry, i'll add you now!

**Comment by ccpandhare** [comment]
Thanks!

**Comment by jywarren** [comment]
:-)

On Mon, Jun 19, 2017 at 12:33 PM, Chinmay Pandhare <notifications@github.com

**Pull request merged**